cluster/cluster.go

```diff
@@ -0,0 +1,472 @@
+	opts         Options
+	knownNodes   node.ServiceNodes
+	usedNodes    idToNodeMap
+	spares       []node.ServiceNode
+	sparesByID   map[string]node.ServiceNode
+	placementSvc services.PlacementService
+	placement    services.ServicePlacement
+	status       Status
+	lastErr      error
+}
+
+// New returns a new cluster backed by provided service nodes
+func New(
+	nodes node.ServiceNodes,
+	opts Options,
+) (Cluster, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	cluster := &svcCluster{
+		logger:       opts.InstrumentOptions().Logger(),
+		opts:         opts,
+		knownNodes:   nodes,
+		usedNodes:    make(idToNodeMap, len(nodes)),
+		spares:       make([]node.ServiceNode, 0, len(nodes)),
+		sparesByID:   make(map[string]node.ServiceNode, len(nodes)),
+		placementSvc: opts.PlacementService(),
+		status:       ClusterStatusUninitialized,
+	}
+	cluster.addSparesWithLock(nodes)
+
+	return cluster, nil
+}
+
+func (c *svcCluster) addSparesWithLock(spares []node.ServiceNode) {
+	for _, spare := range spares {
+		c.spares = append(c.spares, spare)
+		c.sparesByID[spare.ID()] = spare
+	}
+}
+
+func nodeSliceWithoutID(originalSlice node.ServiceNodes, removeID string) node.ServiceNodes {
+	newSlice := make(node.ServiceNodes, 0, len(originalSlice))
+	for _, elem := range originalSlice {
+		if elem.ID() != removeID {
+			newSlice = append(newSlice, elem)
+		}
+	}
+	return newSlice
+}
+
+// TODO(prateek): reset initial placement after teardown
+
+func (c *svcCluster) newExecutor(
+	nodes node.ServiceNodes,
+	fn node.ServiceNodeFn,
+) node.ConcurrentExecutor {
+	return node.NewConcurrentExecutor(nodes, c.opts.NodeConcurrency(), c.opts.NodeOperationTimeout(), fn)
+}
+
+func (c *svcCluster) Placement() services.ServicePlacement {
+	c.Lock()
+	defer c.Unlock()
+	return c.placement
+}
+
+func (c *svcCluster) initWithLock() error {
+	psvc := c.placementSvc
+	_, _, err := psvc.Placement()
+	if err != nil { // attempt to retrieve current placement
+		c.logger.Infof("unable to retrieve existing placement, skipping delete attempt")
+	} else {
+		// delete existing placement
+		err = c.opts.PlacementServiceRetrier().Attempt(psvc.Delete)
+		if err != nil {
+			return fmt.Errorf("unable to delete existing placement during setup(): %+v", err)
+		}
+		c.logger.Infof("successfully deleted existing placement")
+	}
+
+	var (
+		svcBuild        = c.opts.ServiceBuild()
+		svcConf         = c.opts.ServiceConfig()
+		sessionToken    = c.opts.SessionToken()
+		sessionOverride = c.opts.SessionOverride()
+		listener        = c.opts.NodeListener()
+	)
+
+	// setup all known service nodes with build, config
+	executor := c.newExecutor(c.knownNodes, func(node node.ServiceNode) error {
+		err := node.Setup(svcBuild, svcConf, sessionToken, sessionOverride)
+		if err != nil {
+			return err
+		}
+		if listener != nil {
+			node.RegisterListener(listener)
+			// TODO(prateek): track listenerID returned in cluster struct, cleanup in Teardown()
```

Comment:
`node.Teardown()` [clears](https://github.com/m3db/m3em/blob/192163ae86c95a1a0caf5b81611547e51263dd17/node/node.go#L305-L306) listeners, so this isn't strictly needed any more. Will update comment to reflect this.